### PR TITLE
feature(dcd_dwc2) : Added cache synchronization on esp32p4 while DMA is used

### PR DESCRIPTION
**Pull request by roma-jam**
## Requirements
1. To enable Buffer DMA mode, esp-tinyusb changes should be done (define `CFG_TUD_DWC2_DMA` and implement `CFG_TUSB_MEM_SECTION`, refer https://github.com/espressif/esp-usb/pull/88)
2. Support Buffer DMA mode for ESP32S2/S3 (no sync required, so this could be achieved with step 1.) and ESP32P4 (sync required, changes in this PR is crucial for Buffer DMA on P4) chips. 

## Measurements 

Measurements were made on `tusb_msc` example (read-write operations benchmark, SDMMC target).

### ESP32S3

| MSC Buffer size (bytes)   |      Slave Mode (R/W)      |  Buffer DMA (R/W) |
|:----------:|:-------------:|:------:|
| 512 |  463,5 kB/s / 161,1 kB/s | 512,6 kB/s / 171,2 kB/s |
| 4096 |  733,5 kB/s  / 452,9 kB/s  | 878,8 kB/s / 469,9 kB/s |
| 8192 |  N/A / N/A  |  N/A / N/A |

### ESP32P4

| MSC Buffer size (bytes)   |      Slave Mode (R/W)      |  Buffer DMA (R/W) |
|:----------:|:-------------:|:------:|
| 512 |  909,6 kB/s / 217,8 kB/s | 942,8 kB/s / 222,6 kB/s |
| 4096 |    932,7 kB/s / 225,2 kB/s  |  3,8 MB/s / 1,1 MB/s |
| 8192* |  934,4 KB/s / 229,4 KB/s |  5,3 MB/s / 1,8 MB/s |

*Benchmark results provided

image

image

## Limitations

Not working: 
- [ ] tusb_ncm (S2/S3, TinyUSB stack overflow during `recv_transfer_datagram_to_glue_logic()`)

Working examples:
- [x] tusb_composite_msc_serial
- [x] tusb_console
- [x] tusb_hid
- [x] tusb_midi
- [x] tusb_msc
- [x] tusb_serial_device (verified tx with len = [1..512])

## Breaking change
_No breaking changes_

## Checklist

- [x] Pull Request name has appropriate format (for example: "fix(dcd_dwc2): Resolved address selection when several phy are present")
- [ ] _Optional:_ README.md updated
- [x] CI passing

## Related issues
_No related issues_

### CODE REVIEW

**Comment by roma-jam** [comment]
@tore-espressif @peter-marcisovsky 
This is a (preliminary version with using cache sync) DMA support for P4. 
There are couple of changes, regarding the MSC and CDC class drivers, so let me know if it is better to move them out to another PR.

DMA requires the specific configuration by esp-tinyusb component, so the link to the changes: https://github.com/espressif/esp-usb/pull/88

File: src/portable/synopsys/dwc2/dcd_dwc2.c [hunk 1]
```diff
@@ -848,6 +883,11 @@ static void handle_epout_dma(uint8_t rhport, uint8_t epnum, dwc2_doepint_t doepi
 
   if (doepint_bm.setup_phase_done) {
     dma_setup_prepare(rhport);
+    // CACHE HINT
+    // When cache is enabled, _setup_packet must have cache line size alignment
```
Comment (tore-espressif) [inline src/portable/synopsys/dwc2/dcd_dwc2.c:870]: I don't understand the terminology used in this PR.

What does it mean 'cache is enabled'? We sure don't want to enable/disable the cache for all RAM access system wide...

File: src/portable/synopsys/dwc2/dwc2_esp32.h [hunk 2]
```diff
@@ -111,6 +123,31 @@ TU_ATTR_ALWAYS_INLINE static inline void dwc2_phy_update(dwc2_regs_t* dwc2, uint
...
+  int flags = ESP_CACHE_MSYNC_FLAG_DIR_M2C;
+  if (addr != NULL && size) {
+    ESP_EARLY_LOGV("ESP32_DWC", "mem to cache sync, addr 0x%"PRIx32", size %d (%s)", (uintptr_t)addr, size);
+    size = (size < DWC2_MEM_CACHE_LINE_SIZE)? DWC2_MEM_CACHE_LINE_SIZE : size;
+    esp_err_t ret = esp_cache_msync(addr, size, flags);
+    assert(ret == ESP_OK);
+  }
+}
+
+#define dsync_c2m(_addr, _size)   dwc2_dcd_sync_cache_to_memory((_addr), (_size))
```
Comment (tore-espressif) [inline src/portable/synopsys/dwc2/dwc2_esp32.h]: Could we use the interface defined here? https://github.com/hathach/tinyusb/blob/master/src/device/dcd.h#L91

File: src/portable/synopsys/dwc2/dwc2_esp32.h [hunk 3]
```diff
@@ -39,6 +39,18 @@
 #include "soc/periph_defs.h"
 #include "soc/usb_wrap_struct.h"
 
+#if TU_CHECK_MCU(OPT_MCU_ESP32P4)
```
Comment (tore-espressif) [inline src/portable/synopsys/dwc2/dwc2_esp32.h]: The idea of `SOC_CACHE_INTERNAL_MEM_VIA_L1CACHE` was that we don't want our code to be target specific and this symbol can be reused for multiple targets.

```suggestion
```

File: src/portable/synopsys/dwc2/dwc2_esp32.h [hunk 4]
```diff
@@ -111,6 +123,31 @@ TU_ATTR_ALWAYS_INLINE static inline void dwc2_phy_update(dwc2_regs_t* dwc2, uint
...
+    ESP_EARLY_LOGV("ESP32_DWC", "cache to mem sync, addr 0x%"PRIx32", size %d", (uintptr_t)addr, size);
+    esp_err_t ret = esp_cache_msync(addr, size, flags);
+    assert(ret == ESP_OK);
+  }
+}
+
+TU_ATTR_ALWAYS_INLINE static inline void dwc2_dcd_sync_memory_to_cache(void *addr, size_t size) {
+  int flags = ESP_CACHE_MSYNC_FLAG_DIR_M2C;
+  if (addr != NULL && size) {
+    ESP_EARLY_LOGV("ESP32_DWC", "mem to cache sync, addr 0x%"PRIx32", size %d (%s)", (uintptr_t)addr, size);
```
Comment (tore-espressif) [inline src/portable/synopsys/dwc2/dwc2_esp32.h]: Not tested but this should be cleaner

```suggestion
    ESP_EARLY_LOGV("ESP32_DWC", "mem to cache sync, addr %"PRIxPTR", size %d (%s)", addr, size);
```

File: src/portable/synopsys/dwc2/dwc2_esp32.h [hunk 5]
```diff
@@ -39,6 +39,18 @@
 #include "soc/periph_defs.h"
 #include "soc/usb_wrap_struct.h"
 
+#if TU_CHECK_MCU(OPT_MCU_ESP32P4)
+#if (CFG_TUD_DWC2_DMA_ENABLE && SOC_CACHE_INTERNAL_MEM_VIA_L1CACHE)
+#include "sdkconfig.h"
+#include "hal/cache_hal.h"
```
Comment (tore-espressif) [inline src/portable/synopsys/dwc2/dwc2_esp32.h]: If possible, please avoid using hal API in code outside of esp-idf (seems not to be used anywhere here)
```suggestion
```

File: src/portable/synopsys/dwc2/dwc2_esp32.h [hunk 6]
```diff
@@ -111,6 +123,31 @@ TU_ATTR_ALWAYS_INLINE static inline void dwc2_phy_update(dwc2_regs_t* dwc2, uint
   // maybe usb_utmi_hal_disable()
 }
 
+#ifdef DWC2_ENABLE_MEM_CACHE
```
Comment (tore-espressif) [inline src/portable/synopsys/dwc2/dwc2_esp32.h]: I'm unsure why we need this symbol. Can we just reuse `SOC_CACHE_INTERNAL_MEM_VIA_L1CACHE`?

**Review by tore-espressif — COMMENTED** [review]
Just a few work-in-progress comments/questions

File: src/portable/synopsys/dwc2/dcd_dwc2.c (hunk 1 above)
Comment (roma-jam) [reply src/portable/synopsys/dwc2/dcd_dwc2.c:870]: It more like about the way, how we want to handle the memory. while DMA is used. 
I thought, that there will be an option to choose between: 
- Use cache and synchronize it explicitly in dcd_dwc2
- Bypass cache and use direct-access addresses

File: src/portable/synopsys/dwc2/dwc2_esp32.h (hunk 5 above)
Comment (roma-jam) [reply src/portable/synopsys/dwc2/dwc2_esp32.h]: Thanks, removed.

File: src/portable/synopsys/dwc2/dwc2_esp32.h (hunk 4 above)
Comment (roma-jam) [reply src/portable/synopsys/dwc2/dwc2_esp32.h]: I removed debug line from here at all.

File: src/portable/synopsys/dwc2/dwc2_esp32.h (hunk 6 above)
Comment (roma-jam) [reply src/portable/synopsys/dwc2/dwc2_esp32.h]: yep, we can.

File: src/portable/synopsys/dwc2/dwc2_esp32.h [hunk 7]
```diff
@@ -39,6 +39,14 @@
 #include "soc/periph_defs.h"
 #include "soc/usb_wrap_struct.h"
 
+#if (CFG_TUD_DWC2_DMA_ENABLE && SOC_CACHE_INTERNAL_MEM_VIA_L1CACHE)
+#include "sdkconfig.h"
+#include "esp_cache.h"
+#include "esp_log.h"
```
Comment (peter-marcisovsky) [inline src/portable/synopsys/dwc2/dwc2_esp32.h:45]: I saw, you had logging in this file previously. Looks like, this header is not needed anymore.

**Comment by roma-jam** [comment]
This PR is not relevant anymore, as the cache support were implemented in upstream based on the partial commit from here (https://github.com/hathach/tinyusb/pull/2877)
The device classes modifications for buffer alignment and memory sections is in the upstream PR: https://github.com/hathach/tinyusb/pull/2884

So far, temporary changes in class device (cdc and mscd) are not significant, as they will be covered by the merging the PR above.

File: src/portable/synopsys/dwc2/dcd_dwc2.c [hunk 8]
```diff
@@ -348,6 +361,11 @@ static void edpt_schedule_packets(uint8_t rhport, const uint8_t epnum, const uin
 
   const bool is_dma = dma_device_enabled(dwc2);
   if(is_dma) {
+    if (dir == TUSB_DIR_IN && total_bytes != 0) {
+      // CACHE HINT
+      // The xfer->buffer has new data for Host, move it to memory for DMA to transfer it
+      dcd_dcache_clean(xfer->buffer, total_bytes);
```
Comment (peter-marcisovsky) [inline src/portable/synopsys/dwc2/dcd_dwc2.c:367]: I can see that, the functions `cdc_cache_clean` and `dcd_dcache_invalidate` definitions are enclosed into `#if` in `dwc2_esp32.h`

Then, `dcd_dcache_invalidate` function call is placed in this file within the same `#if` as it's definition
But the `dcd_dcache_clean` function call is not. Should we need to add a `#if` for `dcd_dcache_clean` call?

**Comment by peter-marcisovsky** [comment]
@roma-jam  sorry for the late review, I see, it closed now..

**Pull request closed without merging**